File: Labs/Lab 6/lab6.py
def canFinish(numCourses, prerequisites):
    # Construct the graph
    graph = defaultdict(list)
    for course, prereq in prerequisites:
        graph[course].append(prereq)
    
    # Function to perform DFS traversal
    def dfs(course, visited):
        if visited[course] == -1:
            return False  # Cycle detected
        if visited[course] == 1:
            return True   # Already visited and processed
        
        visited[course] = -1  # Mark as visiting
        
        # Visit all neighbors (prerequisites) of the current course
        for neighbor in graph[course]:
            if not dfs(neighbor, visited):
                return False
        
        visited[course] = 1  # Mark as visited and processed
        return True
    
    # Initialize visited array
    visited = [0] * numCourses
    
    # Perform DFS for each course
    for course in range(numCourses):
        if not dfs(course, visited):
            return False  # Cycle detected
    
    return True

File: Labs/Lab 6/lab6.java
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int prereq = prerequisite[1];
            graph.putIfAbsent(course, new ArrayList<>());
            graph.get(course).add(prereq);
        }

        int[] visited = new int[numCourses];

        for (int course = 0; course < numCourses; course++) {
            if (!dfs(course, visited, graph))
                return false;
        }

        return true;
    }

    private boolean dfs(int course, int[] visited, Map<Integer, List<Integer>> graph) {
        if (visited[course] == -1)
            return false;
        if (visited[course] == 1)
            return true;

        visited[course] = -1;

        if (graph.containsKey(course)) {
            for (int neighbor : graph.get(course)) {
                if (!dfs(neighbor, visited, graph))
                    return false;
            }
        }

        visited[course] = 1;
        return true;
    }
